File: src/main/java/com/matamercer/microblog/controllers/api/webfinger/WellKnownController.java
package com.matamercer.microblog.controllers.api.webfinger;

import com.matamercer.microblog.errors.NotFoundException;
import com.matamercer.microblog.models.entities.User;
import com.matamercer.microblog.models.repositories.UserRepository;
import com.matamercer.microblog.utilities.EnvironmentUtil;
import lombok.Getter;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.net.UnknownHostException;
import java.util.*;

@RestController
@RequestMapping("/.well-known")
public class WellKnownController {

    @Autowired
    EnvironmentUtil environmentUtil;

    @Autowired
    UserRepository userRepository;

    @GetMapping("/webfinger")
    @ResponseBody
    public JSONResourceDescriptor getItem(@RequestParam("resource") Optional<String> resource) {
        JSONResourceDescriptor jsonResourceDescriptor = new JSONResourceDescriptor();
        if(resource.isPresent()){
            String[] splitResource = (resource.get()).split("[:@]",3);
            String username = splitResource[1];


            User user = userRepository.findByUsername(username);
            if(user == null){
                throw new NotFoundException();
            }

            jsonResourceDescriptor.setSubject(resource.get());
            try {
                jsonResourceDescriptor.setLinks(Arrays.asList(new Link("self", "application/activity+json", environmentUtil.getServerUrl() + "/activitypub/users/" + user.getId() )));
                return jsonResourceDescriptor;
            } catch (UnknownHostException e) {
                e.printStackTrace();
            }

        }else{
            throw new NotFoundException();
        }

        return jsonResourceDescriptor;
    }

    @Getter
    @Setter
    private class JSONResourceDescriptor{
        private String subject;
        private List<Link> links;
        private List<String> aliases;
        private Map<String, String> properties;
    }

    @Getter
    @Setter
    private class Link{
        private String rel;
        private String type;
        private String href;
        private String titles;
        private String properties;

        public Link(String rel, String type, String href) {
            this.rel = rel;
            this.type = type;
            this.href = href;
        }
    }

}

File: src/main/java/com/matamercer/microblog/web/api/webfinger/WellKnownController.kt
package com.matamercer.microblog.web.api.webfinger

import com.matamercer.microblog.exceptions.NotFoundException
import com.matamercer.microblog.models.repositories.UserRepository
import com.matamercer.microblog.utilities.EnvironmentUtil
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController("/.well-known")
class WellKnownController(
    val environmentUtil: EnvironmentUtil,
    var userRepository: UserRepository
) {
    @GetMapping("/webfinger")
    @ResponseBody
    fun getItem(@RequestParam("resource") resource: Optional<String>): JSONResourceDescriptor {
        val jsonResourceDescriptor = JSONResourceDescriptor()
        return if (resource.isPresent) {
            val splitResource = resource.get().split("[:@]".toRegex(), 3).toTypedArray()
            val username = splitResource[1]
            val user = userRepository.findByUsername(username).orElseThrow {
                throw NotFoundException()
            }
            jsonResourceDescriptor.subject = resource.get()
            try {
                jsonResourceDescriptor.links =
                    listOf(
                        Link(
                            "self",
                            "application/activity+json",
                            environmentUtil.serverUrl + "/activitypub/users/" + user.id
                        )
                    )

            } catch (e: Exception) {
                e.printStackTrace()
            }
            jsonResourceDescriptor
        } else {
            throw NotFoundException()
        }
    }

    inner class JSONResourceDescriptor {
        var subject: String? = null
        var links: List<Link>? = null
        var aliases: List<String>? = null
        var properties: Map<String, String>? = null
    }

    inner class Link(private val rel: String, private val type: String, private val href: String) {
        var titles: String? = null
        var properties: String? = null
    }
}